### PR TITLE
e2e library for handling PGP/MIME messages

### PR DESCRIPTION
**Pull request by yonigoogle**
The original library was built by Yan Zhu from yahoo.
I've ported the library to e2e and made a number of changes to the original code.

The library will require some more work before it's finished, but I'd appreciate any advice/feedback at this point.
As of now, the library only supports the building of an outgoing PGP/MIME message.
I'll soon add the parsing of incoming messages.

@diracdeltas

### CODE REVIEW

File: src/javascript/crypto/e2e/openpgp/pgpmime/mimenode.js [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+  // add a newline after the headers (the empty string cell will eventually be
+  // turned into a newline)
+  lines.push('');
+
+  if (this.content_) {
+    if (transferEncoding.value === constants.Mime.BASE64 ||
+        goog.typeOf(this.content_) !== 'string') {
+      // lines are wrapped at 64 chars, per the PEM protocol. Newline characters
+      // will be stripped out of any encrypted content during decryption
+      lines.push(goog.typeOf(this.content_) === 'string' ?
```
Comment (diracdeltas) [inline src/javascript/crypto/e2e/openpgp/pgpmime/mimenode.js]: the indentation in the following lines seems wrong

File: src/javascript/crypto/e2e/openpgp/pgpmime/mimenode.js [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+
+/**
+ * @fileoverview main module for the safemail app.
+ * @author yzhu@yahoo-inc.com (Yan Zhu)
+ */
+
+goog.provide('e2e.openpgp.pgpmime.MimeNode');
+
+goog.require('e2e.openpgp.pgpmime.Constants');
+goog.require('e2e.openpgp.pgpmime.Text');
```
Comment (diracdeltas) [inline src/javascript/crypto/e2e/openpgp/pgpmime/mimenode.js]: `pgpmime.Text` isn't provided anywhere it seems?

File: src/javascript/crypto/e2e/openpgp/pgpmime/mimenode.js (hunk 1 above)
Comment (yonigoogle) [inline src/javascript/crypto/e2e/openpgp/pgpmime/mimenode.js]: It does look a bit funny, though lint approved it. I'll change it though.

File: src/javascript/crypto/e2e/openpgp/pgpmime/mimenode.js (hunk 2 above)
Comment (yonigoogle) [inline src/javascript/crypto/e2e/openpgp/pgpmime/mimenode.js]: That's strange, for some reason it seems like three files weren't included (text.js, text_test.html, text_test.js). text.js is the file that was supposed to provide `pgpmime.Text`.
I'll push again.

File: src/javascript/crypto/e2e/openpgp/pgpmime/mimenode_test.js [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+var TEXT_CONTENT = 'hi this is a\ntext message';
+var BOUNDARY = '--foo';
+var FINAL_MESSAGE = ['Content-Type: multipart/mixed; boundary="--foo"',
+  'Content-Transfer-Encoding: 7bit',
+  '', '----foo', 'Content-Type: text/plain; charset="utf-8"',
+  'Content-Transfer-Encoding: 7bit', '', TEXT_CONTENT, '----foo',
+  'Content-Type: application/octet-stream',
+  'Content-Transfer-Encoding: base64',
+  'Content-Disposition: attachment; filename="example.txt"', '',
+  e2e.openpgp.pgpmime.Text.prettyTextWrap(
```
Comment (diracdeltas) [inline src/javascript/crypto/e2e/openpgp/pgpmime/mimenode_test.js]: I think it's better to put the literal string here, not call `prettyTextWrap`. That way, the expected result doesn't depend on the function working as expected.

File: src/javascript/crypto/e2e/openpgp/pgpmime/pgpmail.js [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+};
+
+
+/**
+ * Builds a plaintext serialized MIME tree for the email.
+ * @param {!e2e.openpgp.pgpmime.types.MailContent} content The plaintext
+ *     content of the email.
+ * @return {string}
+ */
+e2e.openpgp.pgpmime.PgpMail.prototype.buildMimeTree = function(content) {
```
Comment (diracdeltas) [inline src/javascript/crypto/e2e/openpgp/pgpmime/pgpmail.js]: This method doesn't actually depend on any properties of PgpMail, so better to move it into utils. (Especially since you can't call it until an instance of PgpMail is constructed, which isn't possible until you know the encrypted content, which isn't possible until you buildMimeTree, etc.).

File: src/javascript/crypto/e2e/openpgp/pgpmime/pgpmail_test.js [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+  node = null;
+}
+
+
+function testBuildPGPMimeTree() {
+  var finalTree = ['Content-Type: multipart/encrypted;',
+    'protocol="application/pgp-encrypted"; charset="utf-8"; ' +
+        'boundary="--foo"', 'Content-Transfer-Encoding: 7bit',
+    'Subject: test email', 'From: ystoller@google.com',
+    'To: kbsriram@google.com', 'Mime-Version: 1.0', '',
```
Comment (diracdeltas) [inline src/javascript/crypto/e2e/openpgp/pgpmime/pgpmail_test.js]: Maybe add a bunch of recipients to this test case so you can check if line wrapping works as expected. (This is probably going to happen a lot in real life.)

File: src/javascript/crypto/e2e/openpgp/pgpmime/mimenode.js (hunk 1 above)
Comment (diracdeltas) [inline src/javascript/crypto/e2e/openpgp/pgpmime/mimenode.js]: huh, weird. I guess lint knows better than me.

File: src/javascript/crypto/e2e/openpgp/pgpmime/text.js [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * @fileoverview Utility methods for working with free-text and PGP messages.
```
Comment (diracdeltas) [inline src/javascript/crypto/e2e/openpgp/pgpmime/text.js]: Is there a reason not to just put this in pgpmime/utils.js?

Comment (yonigoogle) [inline src/javascript/crypto/e2e/openpgp/pgpmime/text.js]: I was expecting to soon add more methods to e2e.openpgp.pgpmime.Text. If that doesn't end up happening, I'll move it to utils!

File: src/javascript/crypto/e2e/openpgp/pgpmime/pgpmail_test.js (hunk 5 above)
Comment (yonigoogle) [inline src/javascript/crypto/e2e/openpgp/pgpmime/pgpmail_test.js]: Good idea, I'll do that.

File: src/javascript/crypto/e2e/openpgp/pgpmime/mimenode_test.js (hunk 3 above)
Comment (yonigoogle) [inline src/javascript/crypto/e2e/openpgp/pgpmime/mimenode_test.js]: You're right, I'll change that.

**Comment by diracdeltas** [comment]
lgtm other than comments above. BTW, I should be covered under the Google corporate CLA (for Yahoo).

File: src/javascript/crypto/e2e/openpgp/pgpmime/pgpmail.js (hunk 4 above)
Comment (yonigoogle) [inline src/javascript/crypto/e2e/openpgp/pgpmime/pgpmail.js]: That's a good point. I think I'll actually remove the `content` argument and make it depend on the `this.content` property of PgpMail (just like its counterpart buildPGPMimeTree)

Comment (diracdeltas) [inline src/javascript/crypto/e2e/openpgp/pgpmime/pgpmail.js]: In which case, `this.content` in the constructor needs to be the unencrypted content

File: src/javascript/crypto/e2e/openpgp/pgpmime/mimenode.js (hunk 1 above)
Comment (koto) [inline src/javascript/crypto/e2e/openpgp/pgpmime/mimenode.js]: If you mean lines 161/162, they look Googley to me.

File: src/javascript/crypto/e2e/openpgp/pgpmime/mimenode.js [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * @fileoverview main module for the safemail app.
```
Comment (koto) [inline src/javascript/crypto/e2e/openpgp/pgpmime/mimenode.js]: "main module" -> "Main module"

File: src/javascript/crypto/e2e/openpgp/pgpmime/mimenode.js [hunk 8]
```diff
@@ -0,0 +1,198 @@
...
+
+goog.scope(function() {
+var pgpmime = e2e.openpgp.pgpmime;
+var constants = pgpmime.Constants;
+
+
+
+/**
+ * Constructor for a MIME tree node.
+ * @param {{
```
Comment (koto) [inline src/javascript/crypto/e2e/openpgp/pgpmime/mimenode.js]: This should be `@typedef`ed in pgpmime/types.js

File: src/javascript/crypto/e2e/openpgp/pgpmime/mimenode.js [hunk 9]
```diff
@@ -0,0 +1,198 @@
...
+e2e.openpgp.pgpmime.MimeNode = function(options, opt_parent) {
+  this.parent = opt_parent || this;
+  this.multipart_ = options.multipart;
+
+  this.children_ = [];
+  this.header_ = /** @type {e2e.openpgp.pgpmime.types.Header} */ ({});
+  this.content_ = null;
+  this.boundary_ = '';
+
+  // set headers
```
Comment (koto) [inline src/javascript/crypto/e2e/openpgp/pgpmime/mimenode.js]: Comments should be sentences, so they start with capital letter and end with a fullstop.

File: src/javascript/crypto/e2e/openpgp/pgpmime/mimenode.js [hunk 10]
```diff
@@ -0,0 +1,198 @@
...
+  // TODO: Strictly ensure that the boundary value doesn't coincide with
+  //   any string in the email content and headers.
+  this.boundary_ = '---' + goog.string.getRandomString() +
+      Math.floor(Date.now() / 1000).toString();
+};
+
+
+/**
+ * Adds a child to a MIME node.
+ * @param {{
```
Comment (koto) [inline src/javascript/crypto/e2e/openpgp/pgpmime/mimenode.js]: One more reason to typedef it ;)

File: src/javascript/crypto/e2e/openpgp/pgpmime/mimenode.js [hunk 11]
```diff
@@ -0,0 +1,198 @@
...
+  return node;
+};
+
+
+/**
+ * Sets a MIME header.
+ * @param {string} key Name of the header.
+ * @param {string} value Value of the header, possibly including parameters.
+ * @param {Object.<string, string>=} opt_params Optional dict of additional
+ *   parameters
```
Comment (koto) [inline src/javascript/crypto/e2e/openpgp/pgpmime/mimenode.js]: Indent this line 4 spaces, not counting the first space that comes from JSDoc formatting  (i.e. you're missing two spaces).

File: src/javascript/crypto/e2e/openpgp/pgpmime/mimenode.js [hunk 12]
```diff
@@ -0,0 +1,198 @@
...
+ * @private
+ */
+e2e.openpgp.pgpmime.MimeNode.prototype.setHeader_ =
+    function(key, value, opt_params) {
```
Comment (koto) [inline src/javascript/crypto/e2e/openpgp/pgpmime/mimenode.js]: This would fnt into a  single line, otherwise fit as much as possible there.

File: src/javascript/crypto/e2e/openpgp/pgpmime/mimenode.js [hunk 13]
```diff
@@ -0,0 +1,198 @@
...
+      // will be stripped out of any encrypted content during decryption
+      lines.push(goog.typeOf(this.content_) === 'string' ?
+          e2e.openpgp.pgpmime.Text.prettyTextWrap(
+              goog.crypt.base64.encodeString(goog.asserts.assertString(
+              this.content_)), constants.MimeNum.LINE_WRAP,
+              constants.Mime.CRLF) :
+          e2e.openpgp.pgpmime.Text.prettyTextWrap(
+              goog.crypt.base64.encodeByteArray(
+              /** @type {e2e.ByteArray}*/ (this.content_)),
+              constants.MimeNum.LINE_WRAP, constants.Mime.CRLF));
```
Comment (koto) [inline src/javascript/crypto/e2e/openpgp/pgpmime/mimenode.js]: Make this an `if` statement, it spans several lines.

File: src/javascript/crypto/e2e/extension/constants.js [hunk 14]
```diff
@@ -23,6 +23,7 @@ goog.provide('e2e.ext.constants.Actions');
 goog.provide('e2e.ext.constants.BackupCode');
```
Comment (koto) [inline src/javascript/crypto/e2e/extension/constants.js]: Why is that file needed? These namespaces are not used elsewhere (+ it's in the different, `ext` namespace).

File: src/javascript/crypto/e2e/openpgp/pgpmime/mimenode.js [hunk 15]
```diff
@@ -0,0 +1,198 @@
...
+ * Builds an RFC 2822 message from the node and its children.
+ * @return {string}
+ */
+e2e.openpgp.pgpmime.MimeNode.prototype.buildMessage = function() {
+  var lines = [];
+  var transferEncoding = this.header_[constants.Mime.CONTENT_TRANSFER_ENCODING];
+  var contentType = this.header_[constants.Mime.CONTENT_TYPE];
+
+  if (!this.header_[constants.Mime.CONTENT_DISPOSITION] &&
+              this.content_ && goog.typeOf(this.content_) === 'string') {
```
Comment (koto) [inline src/javascript/crypto/e2e/openpgp/pgpmime/mimenode.js]: var contentIsString = goog.typeOf... 

You're checking the type a few times, this way it's easier to read.

File: src/javascript/crypto/e2e/openpgp/pgpmime/mimenode.js [hunk 16]
```diff
@@ -0,0 +1,198 @@
...
+    // sets the charset to utf-8 if the charset hasn't already been specified
+    contentType.params.charset =
+        contentType.params.charset || constants.Mime.UTF8;
+  }
+  if (this.multipart_) {
+    // Multipart messages need to specify a boundary
+    contentType.params.boundary = this.boundary_;
+  }
+
+  // adds all the headers after serialization (transforming object to string)
```
Comment (koto) [inline src/javascript/crypto/e2e/openpgp/pgpmime/mimenode.js]: Please start with a capital letter here and below.

File: src/javascript/crypto/e2e/openpgp/pgpmime/mimenode.js [hunk 17]
```diff
@@ -0,0 +1,198 @@
...
+ * @param {e2e.openpgp.pgpmime.MimeNode=} opt_parent The parent node.
+ * @constructor
+ */
+e2e.openpgp.pgpmime.MimeNode = function(options, opt_parent) {
+  this.parent = opt_parent || this;
+  this.multipart_ = options.multipart;
+
+  this.children_ = [];
+  this.header_ = /** @type {e2e.openpgp.pgpmime.types.Header} */ ({});
+  this.content_ = null;
```
Comment (koto) [inline src/javascript/crypto/e2e/openpgp/pgpmime/mimenode.js]: Please add @types to all those internal variables. /** @private {type} */ will do, although usually we do:
```
/** 
 * Description
 * @private {type}
 */
this.var_ = xxx;
```

File: src/javascript/crypto/e2e/openpgp/pgpmime/mimenode.js [hunk 18]
```diff
@@ -0,0 +1,198 @@
...
+  }
+  goog.object.set(this.header_, key, headerValue);
+};
+
+
+/**
+ * Sets the content.
+ * @param {(string|!e2e.ByteArray)} content The content to set
+ */
+e2e.openpgp.pgpmime.MimeNode.prototype.setContent = function(content) {
```
Comment (koto) [inline src/javascript/crypto/e2e/openpgp/pgpmime/mimenode.js]: `ByteArray` content will fail in buildContent (it's always `typeof string` or `goog.assertString`). In the test the BINARY_CONTENT is a string, and not a byte array.

In general it's better to just accept one type, `ByteArray` being the more generic. we have conversion functions in `e2e` namespace.

File: src/javascript/crypto/e2e/openpgp/pgpmime/pgpmail.js [hunk 19]
```diff
@@ -0,0 +1,142 @@
...
+  rootNode.setContent(this.preamble);
+
+  // Set the required version info
+  // We're currently treating this node as an attachment (by including a
+  // name field), rather than solely containing the version/control information.
+  var versionNode = rootNode.addChild({multipart: false,
+    optionalHeaders: [
+      {name: constants.Mime.CONTENT_TYPE,
+        value: constants.Mime.ENCRYPTED,
+        params: {'charset': 'utf-8', 'name': 'version.asc'}},
```
Comment (koto) [inline src/javascript/crypto/e2e/openpgp/pgpmime/pgpmail.js]: Make version.asc and encrypted.asc constants.

File: src/javascript/crypto/e2e/openpgp/pgpmime/utils.js [hunk 20]
```diff
@@ -0,0 +1,100 @@
...
+  var parts = text.split('; ');
+  var firstPart = parts.shift();
+
+  // Normalize value to lowercase since it's case insensitive
+  var value = goog.string.stripQuotes(firstPart.toLowerCase().trim(), '"');
+
+  var params = {};
+  goog.array.forEach(parts, function(part) {
+    // Ex: 'protocol=application/pgp-encrypted'
+    var paramParts = part.split('=');
```
Comment (koto) [inline src/javascript/crypto/e2e/openpgp/pgpmime/utils.js]: `paramParts = goog.string.splitLimit(part, '=', 2)`. 

Otherwise you wouldn't accept valid `param="a=value=with=equals"`

Later on you need to check if the parameter name and value are correct (i.e. contain only allowed characters). Parameter value allows for more characters (incl. `=`), but only if it's quoted, so you need to have a code branch for this too. See RFC 2045 5.1.

**Comment by koto** [comment]
LGTM modulo comments I've given.

File: src/javascript/crypto/e2e/openpgp/pgpmime/mimenode.js [hunk 21]
```diff
@@ -0,0 +1,198 @@
...
+        if (goog.isDefAndNotNull(header.params)) {
+          this.setHeader_(header.name, header.value, header.params);
+        } else {
+          this.setHeader_(header.name, header.value);
+        }
+      }
+    }, this));
+  }
+
+  this.setBoundary_();
```
Comment (koto) [inline src/javascript/crypto/e2e/openpgp/pgpmime/mimenode.js]: Set it in buildMessage. That way the constructor is cheaper, and at that time you can generate a boundary that does not collide with the content.

File: src/javascript/crypto/e2e/openpgp/pgpmime/pgpmail.js [hunk 22]
```diff
@@ -0,0 +1,142 @@
...
+
+
+/**
+ * Builds a plaintext serialized MIME tree for the email.
+ * @return {string}
+ */
+e2e.openpgp.pgpmime.PgpMail.prototype.buildMimeTree = function() {
+  var rootNode;
+
+  if (!this.content.attachments || this.content.attachments.length === 0) {
```
Comment (diracdeltas) [inline src/javascript/crypto/e2e/openpgp/pgpmime/pgpmail.js]: This function still confuses me because `this.content` is supposed to represent the encrypted mail (according to the comment). But `buildMimeTree` is supposed to be called on unencrypted content.

File: src/javascript/crypto/e2e/openpgp/pgpmime/pgpmail_test.js [hunk 23]
```diff
@@ -0,0 +1,86 @@
...
+}
+
+
+function tearDown() {
+  stubs.reset();
+  mockControl.$tearDown();
+  node = null;
+}
+
+
```
Comment (diracdeltas) [inline src/javascript/crypto/e2e/openpgp/pgpmime/pgpmail_test.js]: Needs a test for `buildMimeTree`.

File: src/javascript/crypto/e2e/extension/constants.js (hunk 14 above)
Comment (yonigoogle) [inline src/javascript/crypto/e2e/extension/constants.js]: The file isn't needed - I'm not sure why it's being shown here, my branch isn't tracking it. 
I'll look into this.

File: src/javascript/crypto/e2e/openpgp/pgpmime/utils.js (hunk 20 above)
Comment (yonigoogle) [inline src/javascript/crypto/e2e/openpgp/pgpmime/utils.js]: Good point. I'll add code to verify that parameter names and values are valid according to the RFC specification

File: src/javascript/crypto/e2e/openpgp/pgpmime/pgpmail_test.js (hunk 23 above)
Comment (yonigoogle) [inline src/javascript/crypto/e2e/openpgp/pgpmime/pgpmail_test.js]: Added one. I'll push it soon.

File: src/javascript/crypto/e2e/openpgp/pgpmime/pgpmail.js (hunk 22 above)
Comment (yonigoogle) [inline src/javascript/crypto/e2e/openpgp/pgpmime/pgpmail.js]: I've changed the definition of the function. It can now accept both encrypted and unencrypted content based on whether a MIME message or a PGP/MIME  message is being constructed. 
Does that sound alright?

Comment (diracdeltas) [inline src/javascript/crypto/e2e/openpgp/pgpmime/pgpmail.js]: I think so
